src/main/java/picard/util/StringDistanceUtils.java

```diff
@@ -0,0 +1,252 @@
+        final byte[] maskedReadBases;
+        if (readQualities == null || !anySmaller(readQualities, minimumBaseQuality)) {
+            maskedReadBases = readBases;
+        } else {
+            maskedReadBases = maskSmaller(readBases, readQualities, minimumBaseQuality);
+        }
+        return maskedReadBases;
+    }
+
+    public static int levenshteinDistanceForBarcodes(final byte[] barcodeBases, final byte[] readBases, final byte[] readQualities, final int minimumBaseQuality) {
+        final byte[] maskedReadBases = getBytesMasked(readBases, readQualities, minimumBaseQuality);
+
+        return levenshteinDistance(barcodeBases, maskedReadBases, 10);
+    }
+
+    /**
+     * +++lifted from Commons Lang Text +++
+     * <p>
+     * modified to specific to comparing barcodes to reads. This means
+     * ignoring insertions or deletions in the last position as it would amount to
+     * double counting otherwise.
+     * <p>
+     * Also, in case of passing the threshold this algo has been modified to return
+     * threshold + 1
+     * <p>
+     * This implementation only computes the distance if it's less than or
+     * equal to the threshold value, returning threshold + 1 if it's greater. The
+     * advantage is performance: unbounded distance is O(nm), but a bound of
+     * k allows us to reduce it to O(km) time by only computing a diagonal
+     * stripe of width 2k + 1 of the cost table. It is also possible to use
+     * this to compute the unbounded Levenshtein distance by starting the
+     * threshold at 1 and doubling each time until the distance is found;
+     * this is O(dm), where d is the distance.
+     * <p>
+     * One subtlety comes from needing to ignore entries on the border of
+     * our stripe eg. p[] = |#|#|#|* d[] = *|#|#|#| We must ignore the entry
+     * to the leftRev of the leftmost member We must ignore the entry above the
+     * rightmost member
+     * *
+     * As a concrete example, suppose s is of length 5, t is of length 7,
+     * and our threshold is 1. In this case we're going to walk a stripe of
+     * length 3. The matrix would look like so:
+     * <p>
+     * <pre>
+     *    1 2 3 4 5
+     * 1 |#|#| | | |
+     * 2 |#|#|#| | |
+     * 3 | |#|#|#| |
+     * 4 | | |#|#|#|
+     * 5 | | | |#|#|
+     *
+     *
+     * This implementation decreases memory usage by using two
+     * single-dimensional arrays and swapping them back and forth instead of
+     * allocating an entire n by m matrix. This requires a few minor
+     * changes, such as immediately returning when it's detected that the
+     * stripe has run off the matrix and initially filling the arrays with
+     * large values so that entries we don't compute are ignored.
+     *
+     * See Algorithms on Strings, Trees and Sequences by Dan Gusfield for
+     * some discussion.
+     */
+
+    public static int levenshteinDistance(final byte[] barcode, final byte[] read, final int threshold) {
```

Comment:
Is there a reason you can't use the htsjdk's `StringUtil.levenshteinDistance`

Reply:
yes...because it isn't exactly levenshtein...I should change the name.